t/data_migration_pipeline/main.go

```diff
@@ -52,37 +59,58 @@ func parseDataModel(record []string) *DataModel {
 }
 
 // emitResult emits data models to be written to Spanner
-func emitResult(s beam.Scope, lines beam.PCollection) beam.PCollection {
+func emitResult(ctx context.Context, s beam.Scope, lines beam.PCollection) beam.PCollection {
 	dataModels := beam.ParDo(s, func(line string, emit func(*DataModel)) {
 		reader := csv.NewReader(strings.NewReader(line))
 		csvLine, err := reader.Read()
 		if err != nil {
 			log.Fatalf("Failed to read record: %v", err)
 		}
 		emit(parseDataModel(csvLine))
+		count.Inc(ctx, 1)
 	}, lines)
+
 	return dataModels
 }
 
 func main() {
+	// Handle context cancellation.
+	ctx, cancel := signal.NotifyContext(context.Background(),
+		syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	// Set up metric
+	ctx = metrics.SetPTransformID(ctx, "pipeline")
+
 	flag.Parse()
 	beam.Init()
 
 	// Create the pipeline object and the root scope
 	p, s := beam.NewPipelineWithRoot()
 
-	lines, err := textio.Immediate(s, *input)
+	lines, err := textio.Immediate(s, *flagInput)
 	if err != nil {
-		log.Fatalf("Failed to read %v: %v", *input, err)
+		log.Fatalf("Failed to read %v: %v", *flagInput, err)
 	}
 
 	// Convert each line to a data model
-	dataModels := emitResult(s, lines)
+	dataModels := emitResult(ctx, s, lines)
+
+	// Verify data on dry run mode
+	if *flagDryRun {
+		log.Println("dry run start....")
+		if _, err := beamx.RunWithMetrics(ctx, p); err != nil {
```

Comment:
You should print metrics in both dry run and real run. The only difference is for dry run, we don't do spanner write.

Reply:
got it